src/lib.rs

```diff
@@ -50,36 +55,48 @@ pub fn run(args: Args) {
     };
     let supported_language = args.language.get_language();
     let language = supported_language.language();
-    let query = get_query(&query_source, language);
+    let query = Arc::new(get_query(&query_source, language));
     let capture_index = args.capture_name.as_ref().map_or(0, |capture_name| {
         query
             .capture_index_for_name(capture_name)
             .expect(&format!("Unknown capture name: `{}`", capture_name))
     });
 
-    enumerate_project_files(&*supported_language)
-        .par_iter()
-        .for_each(|project_file_dir_entry| {
-            let mut printer = grep::printer::Standard::new_no_color(io::stdout());
-            let path = project_file_dir_entry.path();
-
-            let matcher = TreeSitterMatcher::new(
-                &query,
-                capture_index,
-                language,
-                args.filter.clone(),
-                args.filter_arg.clone(),
-            );
-
-            SearcherBuilder::new()
-                .multi_line(true)
-                .build()
-                .search_path(&matcher, path, printer.sink_with_path(&matcher, path))
-                .unwrap();
-        })
+    let (sender, receiver) = mpsc::channel::<DirEntry>();
```

Comment:
Ok so the basic "thread architecture" here is:
Spawn a "receiver" thread whose job is to get project file `DirEntry`'s from a channel and for each one add a job to a thread pool (for doing the query-matching for that file)

And in the main thread do a parallelized enumerating of the project files and send each one on the channel as they get enumerated